weave/index/index_sql.py

```diff
@@ -0,0 +1,869 @@
+            pantry path (with _ replacements when necessary).
+        """
+        if not _HAS_PYODBC:
+            raise ImportError("Missing Dependency. The package 'pyodbc' "
+                              "is required to use this class")
+
+        # Check that the required environment variables are set.
+        try:
+            # We want to fail early if these are not set, so we check them here
+            # even though they are not used until we connect.
+            # Pylint thinks this is pointless, so pylint is getting ignored.
+            # pylint: disable=pointless-statement
+            os.environ["MSSQL_HOST"]
+            os.environ["MSSQL_USERNAME"]
+            os.environ["MSSQL_PASSWORD"]
+            # pylint: enable=pointless-statement
+        except KeyError as key_error:
+            raise KeyError("The following environment variables must be set "
+                           "to use this class: MSSQL_HOST, MSSQL_USERNAME, "
+                           "MSSQL_PASSWORD.") from key_error
+
+        self._file_system = file_system
+        self._pantry_path = pantry_path
+
+        self._database_name = kwargs.get("database_name", "weave_db")\
+            .replace("-", "_")
+        d_schema_name = self._pantry_path.replace(os.sep, "_")\
+            .replace("-", "_")
+        self._pantry_schema = kwargs.get("pantry_schema", d_schema_name)
+
+        # Make a connection to tempdb, then create the desired database if it
+        # does not exist. Then connect to the desired database, and create the
+        # tables if they do not exist.
+        self._connect(database_name="tempdb", **kwargs)
+        self._select_db(database_name=self._database_name)
+        self._create_schema()
+        self._create_tables()
+
+    # Create read-only properties for the database and schema names.
+    @property
+    def database_name(self):
+        """The database name of the index."""
+        return self._database_name
+
+    @property
+    def pantry_schema(self):
+        """The schema name of the index."""
+        return self._pantry_schema
+
+    def __del__(self):
+        """Close the connection to the database."""
+        self.cur.close()
+        self.con.close()
+
+    def _connect(self, database_name="weave_db", **kwargs):
+        """Connect to the server, and select the specified DB.
+
+        Parameters
+        ----------
+        database_name: str (default='weave_db')
+            The database to connect to.
+        **driver: str (default='{ODBC Driver 18 for SQL Server}')
+            Driver to be used for the connection. Usually similar to:
+            '{ODBC Driver <18, 17, 13...> for SQL Server}'.
+        **encrypt: bool (default=False)
+            Whether or not to encrypt the database (disabled by default, due
+            to the fact that it requires additional configuration).
+        """
+        driver = kwargs.get("odbc_driver", "{ODBC Driver 18 for SQL Server}")
```

Comment:
why are you getting "odbc_driver" when the kwargs in the docstring just says "driver"

Reply:
Fixed